engine/crates/composition/src/ingest_subgraph/nested_key_fields.rs

```diff
@@ -0,0 +1,47 @@
+use crate::subgraphs::*;
+
+/// See [crate::subgraphs::Keys::nested_key_fields].
+pub(super) fn ingest_nested_key_fields(subgraph_id: SubgraphId, subgraphs: &mut Subgraphs) {
+    subgraphs.with_nested_key_fields(|subgraphs, nested_key_fields| {
+        for definition in subgraphs.walk(subgraph_id).definitions() {
+            for key in definition.entity_keys() {
+                for field in key.fields() {
+                    let Some(field_type) = definition
+                        .find_field(field.field)
+                        .and_then(|field| field.r#type().definition(subgraph_id))
+                    else {
+                        continue;
+                    };
+
+                    for subselection_field in &field.subselection {
+                        ingest_nested_key_fields_rec(field_type, subselection_field, nested_key_fields);
+                    }
+                }
+            }
+        }
+    });
+}
```

Comment:
What about:

```suggestion
pub(super) fn ingest_nested_key_fields(subgraph_id: SubgraphId, subgraphs: &mut Subgraphs) {
    subgraphs.with_nested_key_fields(|subgraphs, nested_key_fields| {
        let keys = subgraphs
            .walk(subgraph_id)
            .definitions()
            .flat_map(|definition| definition.entity_keys());

        for key in keys {
            for field in key.fields() {
                let Some(field_type) = key
                    .parent_definition()
                    .find_field(field.field)
                    .and_then(|field| field.r#type().definition(subgraph_id))
                else {
                    continue;
                };

                for subselection_field in &field.subselection {
                    ingest_nested_key_fields_rec(field_type, subselection_field, nested_key_fields);
                }
            }
        }
    });
}
```

You could make this even better if `KeyWalker::fields` would return an iterator of walkers, and `Selection` could know in which key it's in.

Reply:
`definition` would be out of scope, unless you add a map inside the flat map. Selection is not doable because it's not normalized so you can't have unique identifiers, it was discussed in a previous review with Benjamin.